flink-kubernetes/src/main/java/org/apache/flink/kubernetes/configuration/KubernetesConfigOptions.java

```diff
@@ -137,11 +138,17 @@
 		.withDescription("The cluster-id, which should be no more than 45 characters, is used for identifying " +
 			"a unique Flink cluster. If not set, the client will automatically generate it with a random ID.");
 
+	// The default container image that ties to the exact needed versions of both Flink and Scala.
+	public static final String DEFAULT_CONTAINER_IMAGE = "flink:" + EnvironmentInformation.getVersion() + "-scala_" + EnvironmentInformation.getScalaVersion();
+
 	public static final ConfigOption<String> CONTAINER_IMAGE =
 		key("kubernetes.container.image")
 		.stringType()
-		.defaultValue("flink:latest")
-		.withDescription("Image to use for Flink containers.");
+		.defaultValue(DEFAULT_CONTAINER_IMAGE)
+		.withDescription("Image to use for Flink containers. " +
+			"The specified image MUST be based upon Apache Flink " +
+			EnvironmentInformation.getVersion() + " that uses Scala " + EnvironmentInformation.getScalaVersion() + " . " +
```

Comment:
Sorry, I did not notice this before.
This text goes into the user docs. The Scala version used by the build will be fixed in the user docs.
Therefore, It makes sense to require that the Scala version of the user application and the image must be the same.

btw, when any option is changed, the contract now is to [rebuild the option docs](https://github.com/apache/flink/tree/master/flink-docs#configuration-documentation).
This is why the CI fails.

I also think we do not usually highlight imperatives like `MUST` in user docs for style reasons. It looks a bit like shooting...

Reply:
I also think we do not usually highlight imperatives like MUST in user docs for style reasons. It looks a bit like shooting...

Reply:
Fixed the 'MUST'

Reply:
Thanks. I regenerated the documentation which now has the values of the default and the description with the values as they were on my machine when I generated it. I can make the description  'generic' but even then the default value will show "what I have".
So everytime someone changes any config documentation this default in the documentation will change to what they have.

How do you propose we handle this?

Reply:
This is a good point. The default value, generated like this, does not look like making sense then. I think the ideal solution would be to put a default value which includes only released Flink version or `latest` for snapshot, both without Scala version, mostly for documentation purposes.

The actual image to use in runtime, if this option is not explicitly configured by user and what we can document, can be then the current full image tag: Flink and Scala version which were used to build the running Flink. This requires modification of the code which parses the option value.

Ideally, we should also calculate the concrete latest Flink (e.g. 1.10) for snapshot as we do not have docker image for snapshot version (e.g. 1.11-snapshot). This is to combine it with the concrete Scala version, compiled against. For simplicity, we could ignore Scala version for snapshots for now and just use `latest`, like in docs, as it is mostly for development where we could ensure correct Scala version for testing.

Reply:
If you had problems with Flink/Scala version mismatch, maybe the problem is actually more general. The solution could be to embed some explicit runtime check of the versions used to build the user code and the cluster where the user tries to submit the job to. This way we could fail fast with a descriptive mismatch message or at least log the mismatch. I am not aware whether we have something like this. This also sounds like a separate effort.

Reply:
I was thinking about this and the main thing is handling version difference between the flink runtime and the application flink libraries better.

Now I see two things that can be done to avoid this:
1) Avoid: By setting the right default docker image to use in Kubernetes (this story). I think we should simply have a SNAPSHOT image too. I do this for my own projects too https://hub.docker.com/r/nielsbasjes/yauaa/tags 
2) Block: At runtime check the compatibility between the application and the runtime. I agree this is a separate problem. I propose to make a new issue out of that.

So this change is about the Avoid part, yet it turns out that there are for each Flink version at least 2 distinct versions that are incompatible. For the code this is fine as this is generated for each; the documentation is going to 'flip flop'.

So the real problem is that this default depends on the actual build using the specific tooling version of the developer. The problem is that this has more variations than the version of the project and this is reflected in the documentation.
So really the actual default and the default in the documentation should be different!
How can I do that? Because the table in the documentation should simply show "2.12" as the default scala version; yet the actual code should have the REAL scala version to be used as a default.

Reply:
By doing this default container image without the scala version would let the problem exist I intended to solve. Actually the version should "flip flop" with the scala it was built against, just not in the documentation.

So I thought of a different way to stabilized the version just for the documentation. It is a bit onorthodox so I'm curious to hear what you think of this direction.

The alternative direction I had in mind would be extending the ConfigOption with something like "defaultValueForDocumentation".

Reply:
What is the use case of the image's `ConfigOption#defaultValue` in the user code during runtime? A separate `ConfigOption#defaultValueForDocumentation` would make sense if there is a good reason that the user code needs the actual value with Scala version from `ConfigOption#defaultValue` during runtime. From what I see atm, Flink code does not necessarily need it.

Flink uses `ConfigOption#defaultValue` to generate web documentation which should not "flip flop" but Flink does not have to use `ConfigOption#defaultValue` for the actual default value during runtime. The actual default value can be derived by Flink code from Flink and Scala version (same as `ConfigOption#defaultValue` was implemented now) in runtime if user has not configured it explicitly (`configuration.contains("image.option") == false`).

Reply:
When running on K8s the jobmanager needs the exact image name at runtime to dynamically start new task workers. So the user application code does not need the image name, the Flink infra running on K8s does need it at runtime.

So what you propose is to have the defaultValue be only used by the documentation and the actual runtime default be calculated in a different place? I would vote against this. 
Making the way documentation is generated "obscure" (as I did in my last patch) is less of an issue to me than making the actually used runtime code "obscure".

Reply:
@azagrebin I changed it to the "defaultValueForDocumentation" construct as I mentioned and now it all looks so much cleaner and easier to use.
The documentation now shows a stable flink and scala version without "SNAPSHOT" and the actual default contains the exact value as is needed by the version at hand.

I also did a rebase onto the latest master.

Reply:
`defaultValueForDocumentation` now still differs from the actual runtime value. It can be still confusing that the default image has a concrete Scala version 2.12.

I have another suggestion. We have a bunch of options which have placeholders in the default values, e.g. `MetricOptions#REPORTER_CONFIG_PARAMETER`.
We could do the same here:
`flink-1.11-scala_<SCALA_VERSION>`
explaining, what the `<SCALA_VERSION>` is, in the option description.

Reply:
As for `defaultValueForDocumentation`, I just realised, we already have `org.apache.flink.annotation.docs.Documentation.OverrideDefault` annotation for this. Check e.g. `io.tmp.dirs` option. This looks like what we need.

Reply:
Yes, this annotation is better.
There is one downside: in an annotation text you cannot have any kind of dynamic parameter. So we can not even have the version of the project in there.
So I changed it essentially to a description of the default value similar to what was done in the case of the TMP_DIRS option.